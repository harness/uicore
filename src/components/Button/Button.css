.button {
  &[class*='bp3-button'],
  &[class*='bp3-button']:not([class*='bp3-intent-']) {
    /* Override Blueprint styles */
    border-radius: 4px;
    display: inline-flex;
    font-family: var(--font-family);
    font-size: var(--font-size, var(--font-size-normal));
    height: 32px;
    padding: 5px 10px;
    box-shadow: none;
    background-image: none;
    transition: all 0.25s cubic-bezier(0.39, 0.58, 0.57, 1);
    transition-property: background-color, color, border-color;

    /* No intent/None intent */
    border: 1px solid var(--grey-350);
    color: var(--grey-600);
    background-color: var(--white);

    &.round {
      border-radius: 100px;
      padding-left: 15px;
      padding-right: 15px;
    }

    &.iconOnly:not([disabled]) {
      color: var(--grey-450);
    }

    &.link {
      padding: 0;
      border: none;
      height: auto;
      min-height: auto;
      color: var(--blue-500);
      background-color: transparent;
      vertical-align: unset;

      &:hover,
      &:active,
      &[class*='active'] {
        background: none;
        border: none;
        box-shadow: none;
      }

      &:hover:not([disabled]) {
        text-decoration: underline;
        color: var(--blue-600);
      }
    }

    &:hover {
      border: 1px solid var(--blue-400);
      color: var(--blue-500);
      background-color: var(--blue-100);
      box-shadow: none;

      &.iconOnly:not([disabled]) {
        color: var(--blue-450);
      }
    }

    &:active,
    &[class*='active'] {
      border: 1px solid var(--blue-400);
      color: var(--blue-500);
      background-color: var(--blue-200);
    }

    &[disabled] {
      border: 1px solid var(--grey-350);
      color: var(--grey-600);
      background-color: var(--white);
      opacity: 0.5;

      &.link {
        background-color: none;
        border: none;
      }
    }

    &[class*='minimal'] {
      border: none;
      background: none;

      &:hover {
        background-color: var(--blue-100);
      }

      &:active,
      &[class*='active'] {
        background-color: var(--blue-200);
        color: var(--blue-500);
      }
    }

    /* Primary */
    &[class*='intent-primary'] {
      border: 1px solid var(--blue-500);
      color: var(--white);
      background-color: var(--blue-500);

      &:hover {
        border: 1px solid var(--blue-600);
        background-color: var(--blue-600);
      }

      &:active,
      &[class*='active'] {
        border: 1px solid var(--blue-800);
        background-color: var(--blue-800);
      }

      &[disabled] {
        border: 1px solid var(--blue-500);
        background-color: var(--blue-500);
      }

      &[class*='minimal'] {
        border: none;
        background: none;
        color: var(--blue-500);

        &:hover {
          background-color: var(--blue-100);
        }

        &:active,
        &[class*='active'] {
          background-color: var(--blue-200);
        }
      }
    }

    /* Success */
    &[class*='intent-success'] {
      color: var(--white);
      border: 1px solid var(--green-500);
      background-color: var(--green-500);

      &:hover {
        border: 1px solid var(--green-600);
        background-color: var(--green-600);
      }

      &:active,
      &[class*='active'] {
        border: 1px solid var(--green-800);
        background-color: var(--green-800);
      }

      &[disabled] {
        border: 1px solid var(--green-500);
        background-color: var(--green-500);
      }
    }

    /* Warning */
    &[class*='intent-warning'] {
      color: var(--white);
      border: 1px solid var(--yellow-500);
      background-color: var(--yellow-500);

      &:hover {
        border: 1px solid var(--yellow-600);
        background-color: var(--yellow-600);
      }

      &:active,
      &[class*='active'] {
        border: 1px solid var(--yellow-800);
        background-color: var(--yellow-800);
      }

      &[disabled] {
        border: 1px solid var(--yellow-500);
        background-color: var(--yellow-500);
      }
    }

    /* Danger */
    &[class*='intent-danger'] {
      color: var(--white);
      border: 1px solid var(--red-500);
      background-color: var(--red-500);

      &:hover {
        border: 1px solid var(--red-600);
        background-color: var(--red-600);
      }

      &:active,
      &[class*='active'] {
        border: 1px solid var(--red-800);
        background-color: var(--red-800);
      }

      &[disabled] {
        border: 1px solid var(--red-500);
        background-color: var(--red-500);
      }
    }
  }

  &.with-current-color {
    :global {
      .bp3-icon {
        color: currentColor !important;
      }

      .bp3-spinner .bp3-spinner-head {
        stroke: currentColor;
      }
    }

    /* Make inline icons have the same color as the button color */
    > span {
      svg {
        path,
        circle {
          fill: currentColor !important;
        }

        :not(path) {
          stroke: currentColor !important;
        }
      }
    }
  }

  /* Disabled */
  /* stylelint-disable no-descending-specificity */
  &[class*='bp3-disabled'] {
    svg,
    svg > * {
      cursor: not-allowed;
    }
  }
}
