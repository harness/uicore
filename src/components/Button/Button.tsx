import { AnchorButton, Button as BButton, IButtonProps } from '@blueprintjs/core'
import React, { ElementType, HTMLAttributes, MouseEvent, useState } from 'react'
import { Assign } from 'utility-types'
import { Config } from '../../core/Config'
import { OptionalTooltip } from '../../core/Types'
import { Utils } from '../../core/Utils'
import { useIsMounted } from '../../hooks/useIsMounted'
import { Icon, IconName, IconProps } from '../../icons/Icon'
import { PaddingProps } from '../../styled-props/padding/PaddingProps'
import { omitStyledProps, styledClasses, StyledProps } from '../../styled-props/StyledProps'
import styledClass from '../../styled-props/StyledProps.css'
import css from './Button.css'

export interface ButtonProps
  extends Assign<
      Omit<IButtonProps, 'icon' | 'rightIcon' | 'onClick'>,
      Assign<HTMLAttributes<HTMLButtonElement>, StyledProps>
    >,
    OptionalTooltip {
  /** Left icon */
  icon?: IconName

  /** Right icon */
  rightIcon?: IconName

  /** Optional Icon props */
  iconProps?: Partial<IconProps>

  /** onClick event handler */
  onClick?: (event: MouseEvent) => Promise<void> | void

  /** Link href. If provided, Button rendered as Link */
  href?: string

  /** When true, ignore all styling (useful for special button and link which need custom styling)  */
  noStyling?: boolean

  /** Link target attribute, must go with href */
  target?: '_blank'

  /** Link rel attribute, must go with href */
  rel?: string

  /** Component children */
  children?: React.ReactNode

  /** Make button round */
  round?: boolean
}

export interface LinkProps extends ButtonProps {
  active?: boolean
  withoutHref?: boolean
}

export function Button(props: ButtonProps): React.ReactElement {
  const { icon, rightIcon, round } = props
  const [loading, setLoading] = useState(props.loading === true)
  const isMounted = useIsMounted()

  const onClick = async (event: MouseEvent) => {
    // TODO (tan): Improve loading state when props.onClick() is resolved way too fast
    // that showing loading state causes flickering

    if (loading) {
      return
    }

    if (props.onClick) {
      setLoading(true)

      try {
        await props.onClick(event)
      } finally {
        if (isMounted.current) {
          setLoading(false)
        }
      }
    }
  }

  const Component: ElementType = props.href ? AnchorButton : BButton
  // Set icon size to 12px when there's one with button text
  const iconSize = (icon || rightIcon) && (props.text || props.href) ? 16 : undefined
  // Extra left padding for left icon, for right icon, Blueprint already has proper margin
  const leftIconPadding: PaddingProps | undefined = props.text || props.href ? { right: 'xsmall' } : undefined
  const Tag = (props.href ? 'a' : 'button') as React.ElementType
  const normalizedProps = omitStyledProps(
    props,
    'icon',
    'tooltip',
    'tooltipProps',
    'iconProps',
    'noStyling',
    'withoutHref'
  )
  const button = props.noStyling ? (
    <Tag {...normalizedProps} />
  ) : (
    <Component
      {...normalizedProps}
      loading={loading}
      icon={icon && <Icon name={icon} size={iconSize} padding={leftIconPadding} {...props.iconProps} />}
      rightIcon={rightIcon && <Icon name={rightIcon} size={iconSize} {...props.iconProps} />}
      onClick={onClick}
      className={styledClasses(
        props,
        styledClass.font,
        css.button,
        props.href && !(props.icon || props.rightIcon) && !props.intent ? css.link : '',
        !props.text && !props.intent && !props.href ? css.iconOnly : '',
        round ? css.round : ''
      )}
    />
  )

  return (
    <Utils.WrapOptionalTooltip tooltip={props.tooltip} tooltipProps={props.tooltipProps}>
      {button}
    </Utils.WrapOptionalTooltip>
  )
}

/**
 * @deprecated use Link from react-router-dom
 */
export function Link(props: LinkProps): React.ReactElement {
  const extra: LinkProps = {}

  // Make sure all _blank target has proper rel
  // @see https://github.com/yannickcr/eslint-plugin,-react/blob/master/docs/rules/jsx-no-target-blank.md
  if (/_blank/i.test(props.target || '')) {
    extra.rel = 'noreferrer noopener'
  }

  // This change is made to make Link compatible with routing pattern at Harness
  if (!Config.DISABLE_LINK_REWRITE && props.href?.startsWith('/') && !props.href?.startsWith('/#/')) {
    extra.href = `/#${props.href}`
  }

  if (props.disabled) {
    extra.onClick = Utils.stopEvent
  }

  if (props.withoutHref) {
    extra.href = '#'
    extra.onClick = event => {
      Utils.stopEvent(event)
      props.onClick?.(event)
    }
    extra.elementRef = element => {
      if (props.withoutHref && element) {
        element.href = 'javascript:void()'
      }
    }
  }

  return <Button {...props} {...extra} />
}
