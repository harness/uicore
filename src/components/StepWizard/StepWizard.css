.main {
  &.navBar {
    display: grid;
    grid-template-columns: 320px 1fr;
    overflow: hidden;
    position: relative;
    .navBarList {
      background-color: #267cc9;
      padding: 48px;

      border-right: 1px solid #979797;
      grid-template-rows: repeat(auto-fill, var(--spacing-xxlarge));
      overflow: hidden;
      justify-content: center;

      .navStep {
        display: grid;
        grid-template-columns: 28px 1fr;
        margin-bottom: var(--spacing-6);
        cursor: not-allowed;

        &.activeStep {
          cursor: unset;

          .stepName {
            color: var(--grey-50);
            font-weight: bold;
          }
          .number {
            background-color: var(--grey-50);
          }
        }

        &.completedStep {
          cursor: pointer;
        }

        .stepName {
          color: var(--grey-50);
        }
        &.nestedStep {
          padding-left: var(--spacing-large);
        }
        .number {
          padding: 1px var(--spacing-xsmall);
          margin-right: var(--spacing-small);
          background-color: #79a5cc;
          border-radius: 50%;
          color: #267cc9;
          text-align: center;
          width: 19px;
          height: 19px;
          display: inline-block;
        }
      }
    }
  }
  .stepDetails {
    padding: var(--spacing-9) var(--spacing-10) var(--spacing-10) var(--spacing-10);
    background: var(--form-bg);
  }
}

.header {
  display: grid;
  justify-content: center;
}

.title {
  color: var(--grey-50);
  font-size: large;
  font-weight: 600;
  margin-top: var(--spacing-6);
  margin-bottom: var(--spacing-10);
}

.completedIcon {
  background-color: var(--green-500);
  border-radius: 50%;
  height: var(--spacing-large) !important;
  width: var(--spacing-large) !important;
}
.harnessWatermark {
  position: absolute;
  left: -8.19%;
  right: 78.36%;
  top: 66.12%;
  bottom: -17%;
  width: 400px !important;
  z-index: 1;
  svg {
    opacity: 0.1;
  }
}
