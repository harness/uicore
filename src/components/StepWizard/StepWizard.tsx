import React, { MutableRefObject } from 'react'
import cx from 'classnames'
import css from './StepWizard.css'
import { Icon, IconName, IconProps } from '../../icons/Icon'
import { Text } from '../../components/Text/Text'
import { romanize } from '../../core/Utils'
import { isNil } from 'lodash-es'

interface StepChangeData<SharedObject> {
  prevStep: number
  nextStep: number
  prevStepData: SharedObject
}

export type GotoStepArgs<SharedObject> =
  | {
      stepNumber: number
      stepIdentifier?: never
      prevStepData?: SharedObject
    }
  | {
      stepNumber?: never
      stepIdentifier: string
      prevStepData?: SharedObject
    }
export interface StepWizardProps<SharedObject> {
  icon?: IconName
  iconProps?: Omit<IconProps, 'name'>
  title?: string | JSX.Element
  subtitle?: string | JSX.Element
  children:
    | Array<React.ReactElement<StepProps<SharedObject>> | null>
    | React.ReactElement<StepWizardProps<SharedObject>>
  isNavMode?: boolean
  className?: string
  stepClassName?: string
  navClassName?: string
  onStepChange?: (data: StepChangeData<SharedObject | undefined>) => void
  onCompleteWizard?: (data: SharedObject | undefined) => void
  initialStep?: number
}

export interface StepProps<SharedObject> {
  name?: string | JSX.Element
  identifier?: string
  children?: React.ReactElement<StepWizardProps<SharedObject>> | React.ReactNode
  // These props will be passed by wizard
  prevStepData?: SharedObject
  currentStep?: () => number
  totalSteps?: () => number
  nextStep?: (prevStepData?: SharedObject) => void
  previousStep?: (prevStepData?: SharedObject) => void
  gotoStep?: (args: GotoStepArgs<SharedObject>) => boolean
  firstStep?: (prevStepData?: SharedObject) => void
  lastStep?: (prevStepData?: SharedObject) => void
}

interface StepState<SharedObject> {
  activeStep: number
  prevStep: number
  nestedStepWizard?: Array<{
    wizard?: React.ReactElement<StepWizardProps<SharedObject> | null>
    stepIndex: number
  }>
  prevStepData?: SharedObject
  children?: Array<React.ReactElement<StepProps<SharedObject>>>
  stepNames?: string[]
  totalSteps: number
}

// builds step identifier to step number map
// recursive in nature to support nested wizards
const createStepIdentifierToStepNumberMap = <SharedObject,>(
  steps: StepWizardProps<SharedObject>['children'],
  include: boolean,
  stepIdentifierToStepNumberMap: MutableRefObject<Record<string, number>>,
  currentStepNumber: MutableRefObject<number>
) => {
  if (!steps) {
    return
  }
  React.Children.map(steps, step => {
    if (include) {
      const stepIdentifier = step?.props?.identifier || step?.props?.name
      if (stepIdentifier && typeof stepIdentifier === 'string') {
        currentStepNumber.current++
        stepIdentifierToStepNumberMap.current[stepIdentifier] = currentStepNumber.current
      }
    }
    createStepIdentifierToStepNumberMap(
      step?.props.children as StepWizardProps<SharedObject>['children'],
      step?.type === StepWizard,
      stepIdentifierToStepNumberMap,
      currentStepNumber
    )
  })
}

export function StepWizard<SharedObject = Record<string, unknown>>(
  props: StepWizardProps<SharedObject>
): React.ReactElement {
  const {
    className = '',
    isNavMode = true,
    initialStep = 1,
    children,
    stepClassName = '',
    navClassName = '',
    icon = '',
    iconProps,
    title = '',
    subtitle = ''
  } = props
  const stepIdentifierToStepNumberMap = React.useRef<Record<string, number>>({})
  const currentStepNumber = React.useRef<number>(0)
  React.useEffect(() => {
    stepIdentifierToStepNumberMap.current = {}
    currentStepNumber.current = 0
    createStepIdentifierToStepNumberMap<SharedObject>(
      props.children,
      true,
      stepIdentifierToStepNumberMap,
      currentStepNumber
    )
  }, [props.children])
  const [state, setState] = React.useState<StepState<SharedObject>>({
    activeStep: Array.isArray(children) && (initialStep < 1 || initialStep > children.length) ? 1 : initialStep,
    totalSteps: 0,
    prevStep: -1
  })
  const gotoStep = React.useCallback(
    (args: GotoStepArgs<SharedObject>) => {
      const { stepNumber: stepNumberArg, stepIdentifier, prevStepData } = args
      const stepNumber = stepIdentifier
        ? stepIdentifierToStepNumberMap.current[stepIdentifier]
        : (stepNumberArg as number)
      if (state.activeStep === stepNumber) {
        return true
      }
      const stepData = state.nestedStepWizard?.[state.prevStep]
      const nestedWizard = stepData?.wizard
      if (
        !isNil(nestedWizard) &&
        nestedWizard &&
        nestedWizard.props?.onCompleteWizard &&
        stepData?.stepIndex === React.Children.toArray(nestedWizard.props.children as any).length
      ) {
        nestedWizard.props.onCompleteWizard(prevStepData)
      } else if (
        props.onCompleteWizard &&
        state.totalSteps > 0 &&
        stepNumber > 1 &&
        stepNumber === state.totalSteps + 1
      ) {
        props.onCompleteWizard(prevStepData)
        return true
      } else if (!stepNumber || stepNumber > state.totalSteps || stepNumber < 1) {
        // Not a valid step stepNumber
        return false
      }
      setState(prevState => ({ ...prevState, prevStep: prevState.activeStep, activeStep: stepNumber, prevStepData }))
      return true
    },
    [state.activeStep, state.totalSteps, state.prevStep, state.nestedStepWizard]
  )

  React.useEffect(() => {
    const stepData = state.nestedStepWizard?.[state.prevStep - 1]
    const nestedWizard = stepData?.wizard
    if (!isNil(nestedWizard) && nestedWizard && nestedWizard.props?.onStepChange) {
      const stepIndex = stepData?.stepIndex ?? 1
      const stepDif = state.activeStep - state.prevStep
      nestedWizard.props.onStepChange({
        prevStep: stepIndex,
        nextStep: stepIndex + stepDif,
        prevStepData: state.prevStepData
      })
    } else if (props.onStepChange && state.prevStep !== -1) {
      props.onStepChange({ prevStep: state.prevStep, nextStep: state.activeStep, prevStepData: state.prevStepData })
    }
  }, [state.prevStep, state.activeStep, state.prevStepData])

  const currentStep = React.useCallback(() => {
    return state.activeStep
  }, [state.activeStep])
  const totalSteps = React.useCallback(() => {
    return state.totalSteps
  }, [state.activeStep])
  const nextStep = React.useCallback(
    (prevStepData?: SharedObject) => {
      gotoStep({ stepNumber: state.activeStep + 1, prevStepData })
    },
    [gotoStep]
  )
  const previousStep = React.useCallback(
    (prevStepData?: SharedObject) => {
      gotoStep({ stepNumber: state.activeStep - 1, prevStepData })
    },
    [gotoStep]
  )
  const firstStep = React.useCallback(
    (prevStepData?: SharedObject) => {
      gotoStep({ stepNumber: 1, prevStepData })
    },
    [gotoStep]
  )
  const lastStep = React.useCallback(
    (prevStepData?: SharedObject) => {
      gotoStep({ stepNumber: state.totalSteps, prevStepData })
    },
    [gotoStep]
  )

  React.useLayoutEffect(() => {
    if (Array.isArray(props.children)) {
      const propsChild = React.Children.toArray(props.children) as React.ReactElement[]
      const steps: Array<React.ReactElement<StepProps<SharedObject>>> = []
      const stepNames: string[] = []
      let stepIndex = 0
      const nestedStepWizard: StepState<SharedObject>['nestedStepWizard'] = []
      propsChild.forEach((child, i: number) => {
        if (
          child?.type === StepWizard ||
          (child?.props?.children as React.ReactElement<StepWizardProps<SharedObject>>)?.type === StepWizard
        ) {
          let nestedStepWizardChild = child as React.ReactElement<StepWizardProps<SharedObject>>

          if ((child?.props?.children as React.ReactElement<StepWizardProps<SharedObject>>)?.type === StepWizard) {
            nestedStepWizardChild = child?.props?.children as React.ReactElement<StepWizardProps<SharedObject>>
          }

          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          const nestedChild = React.Children.toArray(
            nestedStepWizardChild.props.children as any
          ) as React.ReactElement[]
          nestedChild.forEach((nested, j: number) => {
            steps.push(nested as React.ReactElement<StepProps<SharedObject>>)
            nestedStepWizard.push({ wizard: nestedStepWizardChild, stepIndex: j + 1 })
            stepNames.push((nested && nested.props && nested.props.name) || `Step ${i + 1}-${j + 1}`)
          })
        } else {
          stepIndex++
          nestedStepWizard.push({ stepIndex })
          steps.push(child as React.ReactElement<StepProps<SharedObject>>)
          stepNames.push((child && child.props && child.props.name) || `Step ${i + 1}`)
        }
      })
      setState(prevState => ({
        ...prevState,
        stepNames,
        totalSteps: stepNames.length,
        children: steps,
        nestedStepWizard
      }))
    }
  }, [children])

  const renderStep = () => (
    <React.Fragment>
      {state.stepNames &&
        state.stepNames.map((stepName, index) => {
          const activeStep = index + 1 === state.activeStep
          const completedSteps = state.activeStep > index + 1
          const isNestedStep = isNil(state.nestedStepWizard?.[index]?.wizard) ? false : true
          const stepIndex = isNestedStep
            ? romanize(state.nestedStepWizard?.[index]?.stepIndex || 0, true)
            : state.nestedStepWizard?.[index]?.stepIndex
          const isNestedFirstStep = isNestedStep ? state.nestedStepWizard?.[index]?.stepIndex === 1 : false
          return (
            <div
              key={index}
              onClick={() => completedSteps && gotoStep({ stepNumber: index + 1, prevStepData: state.prevStepData })}
              className={cx(
                css.navStep,
                navClassName,
                { [css.activeStep]: activeStep },
                { [css.completedStep]: completedSteps },
                { [css.nestedStep]: isNestedStep }
              )}>
              {isNestedFirstStep && state.nestedStepWizard?.[index]?.wizard?.props?.title && (
                <>
                  <div style={{ gridColumn: '1/ span 2', color: 'var(--white)' }}>
                    {state.nestedStepWizard?.[index]?.wizard?.props?.title}
                  </div>
                </>
              )}
              {completedSteps ? (
                <span className={css.completedIcon}>
                  <Icon name="small-tick" size={20} color="grey200" />
                </span>
              ) : (
                <>{typeof stepName === 'string' && <span className={css.number}>{stepIndex}</span>}</>
              )}
              <Text className={css.stepName} lineClamp={2} width={240}>
                {stepName}
              </Text>
              {!isNestedFirstStep && state.nestedStepWizard?.[index]?.stepIndex === 1 ? (
                <Text className={css.stepName} lineClamp={2} width={240}>
                  {subtitle}
                </Text>
              ) : null}
            </div>
          )
        })}
    </React.Fragment>
  )

  const activeChild = state.children?.[state.activeStep - 1] || <div />

  const childProps: StepProps<SharedObject> = {
    currentStep,
    totalSteps,
    prevStepData: state.prevStepData,
    nextStep: nextStep,
    previousStep,
    gotoStep,
    firstStep,
    lastStep
  }

  return (
    <div className={cx(css.main, className, { [css.navBar]: isNavMode })}>
      {isNavMode && (
        <div className={css.navBarList}>
          {icon ? (
            <span className={css.header}>
              <Icon name={icon} {...iconProps} />
            </span>
          ) : null}
          {title ? (
            typeof title === 'string' ? (
              <span className={cx(css.title, css.header)}>{title}</span>
            ) : (
              title
            )
          ) : null}
          {renderStep()}
        </div>
      )}
      {state.activeStep && (
        <div className={cx(css.stepDetails, stepClassName)}> {React.cloneElement(activeChild, childProps)}</div>
      )}
    </div>
  )
}
