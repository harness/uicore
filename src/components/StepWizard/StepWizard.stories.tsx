/* eslint-disable no-console */
import React from 'react'
import type { Meta, Story } from '@storybook/react'
import { ExampleWizard, ExampleStep, ModalExample } from './example/StepWizardExample'
import { Text } from '../Text/Text'
import { StepWizard } from './StepWizard'
import { Color } from '../../core/Color'

export default {
  title: 'Components / Step Wizard',
  component: ExampleWizard
} as Meta

export const Basic: Story<void> = _args => (
  <StepWizard icon="app-kubernetes" iconProps={{ size: 50 }} title="Kubernetes Cluster">
    <ExampleStep name="Create a New Project" />
    <ExampleStep name={`New Project`} />
    <ExampleStep name="Collaborator" />
  </StepWizard>
)

export const FullExample: Story<void> = _args => <ExampleWizard />

export const ModalStepExample: Story<void> = _args => <ModalExample />

export const Nested: Story<void> = _args => (
  <StepWizard
    title="Kubernetes Cluster"
    subtitle={
      <Text
        style={{ marginLeft: 24, marginTop: 'var(--spacing-6)' }}
        icon="service-dockerhub"
        iconProps={{ color: Color.WHITE, size: 26 }}
        color={Color.WHITE}>
        Docker subtitle
      </Text>
    }>
    <ExampleStep name="Artifact Repository Type" />

    <ExampleStep name={`New Project`} />
    <StepWizard
      title="Create new Connector"
      onCompleteWizard={arg => console.log('onCompleteWizard', arg)}
      onStepChange={arg => console.log('Step Change', arg)}>
      <ExampleStep name="Overview" />
      <ExampleStep name="Details" />
      <ExampleStep name="Test Connection" />
    </StepWizard>
    <ExampleStep name="Collaborator" />
    <StepWizard
      onCompleteWizard={arg => console.log('onCompleteWizard', arg)}
      onStepChange={arg => console.log('Step Change', arg)}>
      <ExampleStep
        name={
          <Text icon="service-dockerhub" iconProps={{ color: Color.WHITE, size: 26 }} color={Color.WHITE}>
            Docker Registry
          </Text>
        }
      />
    </StepWizard>
    <ExampleStep name="Final Step" />
  </StepWizard>
)

export const NestedChildSteps: Story<void> = _args => (
  <StepWizard title="Kubernetes Cluster">
    <ExampleStep name="Artifact Repository Type" />
    <div>
      <StepWizard
        onCompleteWizard={arg => console.log('onCompleteWizard', arg)}
        onStepChange={arg => console.log('Step Change', arg)}>
        <ExampleStep name={'Docker Registry'} />
      </StepWizard>
    </div>
    <ExampleStep name={`New Project`} />
    <StepWizard
      title="Create new Connector"
      onCompleteWizard={arg => console.log('onCompleteWizard', arg)}
      onStepChange={arg => console.log('Step Change', arg)}>
      <ExampleStep name="Overview" />
      <ExampleStep name="Details" />
      <ExampleStep name="Test Connection" />
    </StepWizard>
    <ExampleStep name="Collaborator" />
  </StepWizard>
)
