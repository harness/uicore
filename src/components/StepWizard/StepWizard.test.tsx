import React from 'react'
import { StepWizard, StepProps } from './StepWizard'
import { Layout } from '../../layouts/Layout'
import { Button } from '../Button/Button'
import { render, fireEvent, wait, queryByText, waitFor } from '@testing-library/react'

interface SharedObject {
  prevStepName: string | JSX.Element
}

const ExampleStep: React.FC<StepProps<SharedObject>> = props => {
  const currentStep = props.currentStep && props.currentStep()
  return (
    <React.Fragment>
      <Layout.Vertical>
        <div style={{ height: '180px' }}>
          {props.name}: {currentStep}
          <div>Last Step: {props.prevStepData && props.prevStepData.prevStepName}</div>
        </div>
        <div style={{ height: '40px' }}>
          <Button
            disabled={currentStep === 1}
            onClick={() => props.previousStep && props.previousStep({ prevStepName: props.name || '' })}
            text="Previous"
          />
          <Button
            intent="primary"
            style={{ float: 'right' }}
            onClick={() => props.nextStep && props.nextStep({ prevStepName: props.name || '' })}
            text="Next"
          />
        </div>
      </Layout.Vertical>
    </React.Fragment>
  )
}

interface ExampleProps {
  initialStep: number
  stepChange: (data: any) => void
  onSubmit: (data: any) => void
}

const ExampleWizard = ({ initialStep, stepChange, onSubmit }: ExampleProps) => {
  const [counter, setCounter] = React.useState(1)
  return (
    <div>
      Steps Changed: {counter}
      <StepWizard
        icon="app-kubernetes"
        iconProps={{ size: 37 }}
        title="Kubernetes Cluster"
        initialStep={initialStep}
        onStepChange={values => {
          setCounter(counter + 1)
          stepChange(values)
        }}
        onCompleteWizard={onSubmit}>
        <ExampleStep name="Create a New Project" />
        <ExampleStep name={`New Project ${counter}`} />
        <ExampleStep name="Collaborator" />
      </StepWizard>
    </div>
  )
}

const getDefaultProps = (initialStep: number) => ({ initialStep, stepChange: jest.fn(), onSubmit: jest.fn() })

describe('REnder basic Step Wizard', () => {
  test('should render with initial step 1', async () => {
    const { container } = render(<ExampleWizard {...getDefaultProps(0)} />)
    await waitFor(() => queryByText(container as HTMLElement, /Create a New Project: 1/))
    expect(container).toMatchSnapshot()
  })
  test('should render with initial step -1, render step 1', async () => {
    const { container } = render(<ExampleWizard {...getDefaultProps(-1)} />)
    await waitFor(() => queryByText(container as HTMLElement, /Create a New Project: 1/))
    expect(queryByText(container as HTMLElement, /Create a New Project: 1/)).toBeDefined()
  })
  test('should render with initial step 200, render step 1', async () => {
    const { container } = render(<ExampleWizard {...getDefaultProps(200)} />)
    await waitFor(() => queryByText(container as HTMLElement, /Create a New Project: 1/))
    expect(queryByText(container as HTMLElement, /Create a New Project: 1/)).toBeDefined()
  })

  test('should render with initial step 2, render Step 2', async () => {
    const { container } = render(<ExampleWizard {...getDefaultProps(2)} />)
    await waitFor(() => queryByText(container as HTMLElement, /New Project: 2/))
    expect(queryByText(container as HTMLElement, /New Project: 2/)).toBeDefined()
  })

  test('should test Prev and Next Step', async () => {
    const props = getDefaultProps(1)
    const { container } = render(<ExampleWizard {...props} />)
    await waitFor(() => queryByText(container as HTMLElement, /Next/))
    fireEvent.click(queryByText(container as HTMLElement, /Next/) as HTMLDivElement)
    await waitFor(() => queryByText(container as HTMLElement, /New Project: 2/))
    expect(props.stepChange).toHaveBeenLastCalledWith({
      nextStep: 2,
      prevStep: 1,
      prevStepData: {
        prevStepName: 'Create a New Project'
      }
    })
    fireEvent.click(queryByText(container as HTMLElement, /Next/) as HTMLDivElement)
    await waitFor(() => queryByText(container as HTMLElement, /New Project 3/) as HTMLDivElement)
    expect(queryByText(container as HTMLElement, /Collaborator: 3/)).not.toBeNull()
    expect(queryByText(container as HTMLElement, /Last Step: New Project 2/)).not.toBeNull()

    expect(props.stepChange).toHaveBeenLastCalledWith({
      nextStep: 3,
      prevStep: 2,
      prevStepData: {
        prevStepName: 'New Project 2'
      }
    })

    fireEvent.click(queryByText(container as HTMLElement, /Next/) as HTMLDivElement)

    await wait()

    expect(props.onSubmit).toHaveBeenLastCalledWith({
      prevStepName: 'Collaborator'
    })

    fireEvent.click(queryByText(container as HTMLElement, /Previous/) as HTMLDivElement)

    expect(props.stepChange).toHaveBeenLastCalledWith({
      nextStep: 2,
      prevStep: 3,
      prevStepData: {
        prevStepName: 'Collaborator'
      }
    })
    fireEvent.click(queryByText(container as HTMLElement, /Previous/) as HTMLDivElement)

    expect(props.stepChange).toHaveBeenLastCalledWith({
      nextStep: 1,
      prevStep: 2,
      prevStepData: {
        prevStepName: 'New Project 4'
      }
    })
    expect(queryByText(container as HTMLElement, /Create a New Project: 1/)).not.toBeNull()
    expect(queryByText(container as HTMLElement, /Last Step: New Project 4/)).not.toBeNull()
  })
})
