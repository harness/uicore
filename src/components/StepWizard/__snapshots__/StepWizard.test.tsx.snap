// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`REnder basic Step Wizard should render with initial step 1 1`] = `
<div>
  <div>
    Steps Changed: 
    1
    <div
      class="main navBar"
    >
      <div
        class="navBarList"
      >
        <span
          class="header"
        >
          <span
            class="bp3-icon main"
            data-icon="app-kubernetes"
          >
            <svg
              height="37"
              width="37"
            />
          </span>
        </span>
        <span
          class="title header"
        >
          Kubernetes Cluster
        </span>
        <div
          class="navStep activeStep"
        >
          <span
            class="number"
          >
            1
          </span>
          <p
            class="font lineclamp main stepName width width-200"
            style="--text-line-clamp: 2; width: 200px;"
          >
            Create a New Project
          </p>
          <p
            class="font lineclamp main stepName width width-200"
            style="--text-line-clamp: 2; width: 200px;"
          >
            
          </p>
        </div>
        <div
          class="navStep"
        >
          <span
            class="number"
          >
            2
          </span>
          <p
            class="font lineclamp main stepName width width-200"
            style="--text-line-clamp: 2; width: 200px;"
          >
            New Project 1
          </p>
        </div>
        <div
          class="navStep"
        >
          <span
            class="number"
          >
            3
          </span>
          <p
            class="font lineclamp main stepName width width-200"
            style="--text-line-clamp: 2; width: 200px;"
          >
            Collaborator
          </p>
        </div>
      </div>
      <div
        class="stepDetails"
      >
         
        <div
          class="vertical layout-spacing-undefined main"
        >
          <div
            style="height: 180px;"
          >
            Create a New Project
            : 
            1
            <div>
              Last Step: 
            </div>
          </div>
          <div
            style="height: 40px;"
          >
            <button
              class="bp3-button bp3-disabled button font main with-current-color"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <span
                class="bp3-button-text"
              >
                Previous
              </span>
            </button>
            <button
              class="bp3-button button font main intent intent-primary style style-float style-float-right with-current-color"
              style="float: right;"
              type="button"
            >
              <span
                class="bp3-button-text"
              >
                Next
              </span>
            </button>
          </div>
        </div>
      </div>
      <span
        class="bp3-icon main harnessWatermark color color-grey50"
        data-icon="harness-with-color"
      >
        <svg
          height="346"
          width="346"
        />
      </span>
    </div>
  </div>
</div>
`;
